### ISSUE TITLE
Add acoustic echo cancelation

### ISSUE THREAD

**Issue by agilob** [qTox/qTox#1655]

**Comment by tux3** [comment]
So, what would be the difference with the current filter_audio?

**Comment by agilob** [comment]
When two people use loudspeakers (or use echobot on loudspeakers) microphone records sound from speakers and transmits it. Echo cancellation should prevent recording what is played from speakers.

**Comment by tux3** [comment]
But isn't that exactly what filter_audio is supposed to do?

**Comment by agilob** [comment]
Than it doesn't work. For me audio filter removes nothing more than 'white noise'.

**Comment by tux3** [comment]
Yes, that seems to be a bug in filter_audio, the purpose of filter_audio was echo cancellation. Shouldn't we try to fix filter_audio instead of creating a new one?

**Comment by agilob** [comment]
I didn't know what audio filter was supposed to do, I apparently doesn't work with PA + alsa. There are [over 40 issues](https://github.com/tux3/qTox/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+audio) with audio in title :/

**Comment by tux3** [comment]
Yes, it sounds like it does do _something_ with the audio but not anything useful at the moment.

**Comment by An-dy** [comment]
Can we introduce prioritisation voting ?
I would love to vote for features/bugs that bother me and my peers when we use qtox and utox on a daily base.

**Comment by tux3** [comment]
@An-dy let's keep the discussion in one place : #1688

**Comment by zetok** [comment]
That's basically what filter_audio is supposed to do. I think that lately it got better (still not perfect though). Might want to test things again?

**Comment by vjay82** [comment]
I tested this today with the current release 1.2.4.307 on OS X. 
I used 2 Laptops in different rooms. Audio Filtering was enabled in qTox. I called the other laptop and  accepted the call there and immediately got the loudspeakers blasted out by the feedback loop.
What exactly does the Audio Filter checkbox stand for? 
Sadly this way qTox is still no alternative to Skype :(

**Comment by ProMcTagonist** [comment]
isn't filter_audio @irungentoo's?

**Comment by bignaux** [comment]
try with pulseaudio : https://github.com/tux3/qTox/wiki/Pulseaudio#noise-and-echo-cancellation--module-echo-cancel

**Comment by Plinz** [comment]
It could be a good solution to use a free package like Oslec or Speex ?
Because for me the soft is useless, I can't have a discussion without a horrible echo.

**Comment by antis81** [comment]
@Plinz Echo should be better controllable with latest version - we recently implemented a gain amplifier ranging from -30 to 30dB. Go below the centre marker in audio settings and echo should become less for you. We're planning to reimplement audio stuff on Opus (aka libopus). It is already available in Tox, but not for qTox audio yet.

**Comment by Plinz** [comment]
@antis81 Thanks for the update ! I will retry later :)

**Comment by drrmmng** [comment]
Any news? :)

**Comment by beekeeper1** [comment]
Qtox is good but not perfect yet, the echo cancellation still needs a bit of work, one other thing which would be a vast improvement would be the answer call button to be always on top, i find i have to click on a contacts name before i can answer the call as the call button is hiding, sadly i am no good with linux programming and too old to learn now, my knowledge  extends to visual basic and commodore o/s

**Comment by tele1** [comment]
+1 Add to QTox sound filter OPUS or other

**Comment by ghost** [comment]
+1 for echo limiter, echo cancellation

**Comment by svedbox** [comment]
+1 add to qTox echo cancellation

**Comment by baimafeima** [comment]
Uh, so this issue was raised on May 15, 2015. Has there been any progress in getting this fixed? I am running qtox on Solus and as of now, voice and video chat conversations are impossible due to the heavy echoes involved. I've raised an issue on the Solus development tracker as well: https://dev.solus-project.com/T4502

Alternatively, please instruct whether there are any options to _temporarily_ fix this by making changes to either qtox packaging (by distribution maintainers) or changes in the programme settings (by end users).

**Comment by Diadlo** [comment]
@baimafeima You can test echo cancellation feature from current master (Settings - Audio - Enable experimental audio backend)

**Comment by ArisB52** [comment]
@Diadlo I can't find the option for the experimental audio backend, has it been removed?
Tox v: v1.11.0-162-gce0c1020
core v: 0.1.10

Edit:
Nevermind, had to install filter_audio for the checkbox to appear. It does not cancel out the echo btw :)

Edit2: I got it working using:
PULSE_PROP="filter.want=echo-cancel" ./qTox

**Comment by sudden6** [comment]
@ArisB52 can you describe how exactly you tested the echo canceling built into qTox? Maybe we can improve it then.

**Comment by ArisB52** [comment]
@sudden6 Sure, I am on Fedora 26 and installed qTox using git (now 1.12). My friend is on Gentoo and got qTox from emerge (version 1.11). I have a headset, my friend has speakers and a mic. When we are audio-calling I can hear myself talking with a small delay. I have installed filter_audio. Now with the experimental audio backend enabled the overal sound quality degrades. I hear slight beeps en crackling sounds and also myself.
We also tried mumble, mumble apparently adds an echo cancelled playback/capture device which seems to work.
Using the instructions here: https://www.reddit.com/r/linux/comments/2yqfqp/just_found_that_pulseaudio_have_noise/
I now have that echo canceling device available in qTox too and it works quite well (experimental audio disabled)

**Comment by sudden6** [comment]
@ArisB52 uhm, if I get this right, your friend is the one that needs to run the qTox version with echo cancelling, since he is creating the echo with the mic+speakers combination.

As far as echo cancellng via pulseaudio goes, it's also documented for qTox [here](https://github.com/qTox/qTox/wiki/Pulseaudio), but unfortunately it's not a general solution for all supported platforms.

**Comment by tbertels** [comment]
The low priority given to this issue is surprising given the fact that most users (those not using headphones) won't use qTox unless it's fixed.

**Comment by sudden6** [comment]
@tbertels There is experimental support for this on Linux, but it's not very stable. Unfortunately there are no good cross platform libraries to solve this problem.

**Comment by tbertels** [comment]
Have you tried https://github.com/BelledonneCommunications/mediastreamer2?
See also http://www.linphone.org/technical-corner/mediastreamer2/overview

**Comment by sudden6** [comment]
No, but after a quick glance it does much more than removing echo. It could be an example on how to use echo cancellation with the speex library correctly.

**Comment by tbertels** [comment]
You may want to test Linphone first to see if it works there though, because my tests with Linphone to SIP router to DECT are not good, the sound isn't understandable on either side.

**Comment by tele1** [comment]
**@sudden6** It is true. Linux don't have Equalizer for pulse_audio and alsa with plugins.
E.g when I buy computer I have Equalizer on CD for my music card, but this Equalizer is not for Linux. :disappointed: 

Can you try talk with developers **Audacity** , **Realtek** , **Alsa** , **Pulse Audio**
 for create Equalizer for Linux or for QTox ?

For example inside Audacity it looks like this ( Remove Noise in Audacity )
https://www.youtube.com/watch?v=XClduWGEdq0

Similar procedure to remove noise exists in **Team Speak** at the first start.
You can test and talk with this developers, maybe they also can help.

**Edit**
Check also this ( Remove reverb in Audacity )
https://www.youtube.com/watch?v=DJpprNccGNA

**Edit**
Check also this ( Remove Room Tone from your Audio )
https://www.youtube.com/watch?v=xbXpNTvDQrc

**Edit**
**Audacity** have GVerb Reverb plugins, maybe developers can help create Equalizer for reduce reverb and echo effect ?
https://ttmanual.audacityteam.org/man/GVerb
https://manual.audacityteam.org/man/reverb.html

**Comment by anthonybilinski** [comment]
A couple more ideas, Rnnoise looks promising, though the one lib I looked at was not compatible (BSD 3-clause), the algorithm is open though.

And from https://github.com/qTox/qTox/issues/6269#issuecomment-762640720, `VoiceEngine` is recommended saying it includes acoustic echo cancellation (AEC) and noise suppression (NS).

**Comment by ghost** [comment]
+1 add to qTox echo cancellation
Really hard to talk..
Any progress on this so far?

(btw, haven't seen any experimental audio stuff in settings in the last 1.17.3 on Windows)
